fix: use constants

tibor-reiss opened the pull request:
Partially fixes #117

Todo:

- [ ] anthropic: WRAPPED_METHODS -> span_name and metric_name: are always the same. This does not seem correct, is it?
**You're right. Metric name should be removed because the exception count should just be an attribute of the latency metric. I've fixed the span name.**
- [ ] weaviate: **Namespace shouldn't use the Span Attributes constant. The Span Attributes constant should only be used for actual attributes we put on spans.**


Manual testing due to changes to opentelemetry_semantic_conventions_ai
- [x] anthropic
- [x] bedrock
- [x] chromadb
- [x] cohere
- [x] langchain
- [x] llamaindex
- [x] milvus
- [x] mistralai
- [x] ollama
- [x] openai
- [x] pinecone
- [x] qdrant
- [x] replicate
- [x] transformers
- [x] vertexai
- [x] watsonx
- [x] weaviate
- [x] semantic-conventions-ai
- [x] traceloop-sdk

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/traceloop/openllmetry?pullRequest=1131) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/traceloop/openllmetry?pullRequest=1131) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/traceloop/openllmetry?pullRequest=1131) it.</sub>

tibor-reiss commented:
@nirga Is there an easy way to install the packages in editable mode? This PR touches several packages, e.g. opentelemtry-semantic-conventions-ai, which is a dependency for the others. However, the version is baked into the pyproject.toml files. For now, I built the wheels manually and also manually installed them into the respective .venv - but this is very cumbersome. Is there a better way, please?

nirga commented:
This looks good btw @tibor-reiss but we have more :) it's basically every string we have that is a metric name or an attribute should be a semantic convention

tibor-reiss commented:
@nirga Yeah, I saw that :) Just didn't want to make a too large commit before getting any feedback. Feel free to merge this. I'll work on more asap and will add it to this PR (if not merged) or to a new one.

tibor-reiss commented:
@nirga What about env variables / context values, shall those be added also to semantic-conventions? e.g. `def should_send_prompts():` has both.

nirga commented:
Those are up for a different refactor actually @tibor-reiss :) We need to replace them with configuration parameters in each instrumentation's `__init__`, and then pass the environment parameter from the `traceloop_sdk` (or allow configuring it from the `Traceloop.init` method

tibor-reiss commented:
@nirga I have tested it manually by installing semantic-conventions-ai manually into every venv. Since it will not pass on the CI, how would you like to proceed? One option is to make a separate PR for just semantic-conventions-ai and merge that first, because there are only additions to the respective files.

nirga commented:
@tibor-reiss once you add everything to the semconv I'll manually publish a package and update this PR which will make it pass

tibor-reiss commented:
@nirga Tests all completed with the new version. Tried some sample apps (e.g. weaviate) as well. Please check the PR and let me know if anything.

nirga commented on packages/opentelemetry-instrumentation-milvus/opentelemetry/instrumentation/milvus/wrapper.py:
```diff
@@ -92,91 +93,91 @@ def count_or_none(obj):
 def _set_insert_attributes(span, kwargs):
     _set_span_attribute(
-        span, "db.milvus.insert.collection_name", kwargs.get("collection_name")
+        span, f"{AISpanAttributes.MILVUS_INSERT}.collection_name", kwargs.get("collection_name")
```
This also needs to be a constant. So there should be something in `SpanAttributes` that's called
`MILVUS_INSERT_COLLECTION_NAME` that is used here. The reason we need it is because there are observability system that automatically generates dashboards etc. from these constants so we need to be explicit about them.

nirga commented on packages/opentelemetry-instrumentation-openai/tests/traces/test_assistant.py:
```diff
@@ -47,26 +49,26 @@ def test_new_assistant(exporter, openai_client, assistant):
+    assert open_ai_span.attributes[SpanAttributes.LLM_RESPONSE_MODEL] == "gpt-4-turbo-preview"
     assert (
-        open_ai_span.attributes["gen_ai.prompt.0.content"]
+        open_ai_span.attributes[f"{SpanAttributes.LLM_PROMPTS}.0.content"]
```
This is ok, just because this will be refactored into events pretty soon so no need to change it and the likes of it

nirga commented on packages/opentelemetry-instrumentation-pinecone/opentelemetry/instrumentation/pinecone/query_handlers.py:
```diff
@@ -18,21 +18,21 @@ def set_query_input_attributes(span, kwargs):
-    set_span_attribute(span, "pinecone.query.queries", kwargs.get("queries"))
-    set_span_attribute(span, "pinecone.query.top_k", kwargs.get("top_k"))
-    set_span_attribute(span, "pinecone.query.namespace", kwargs.get("namespace"))
+    set_span_attribute(span, f"{SpanAttributes.PINECONE_QUERY}.id", kwargs.get("id"))
```
This needs to be extracted to `SpanAttributes.PINECONE_QUERY_ID`

nirga commented on packages/opentelemetry-instrumentation-pinecone/tests/test_query.py:
```diff
@@ -152,12 +154,12 @@ def test_pinecone_retrieval(traces_exporter, metrics_reader, openai_client):
             for metric in sm.metrics:
 
-                if metric.name == "db.pinecone.query.duration":
+                if metric.name == f"{SpanAttributes.PINECONE_DB_QUERY}.duration":
```
`Meters.VECTOR_DB_OPERATION_DURATION`

nirga commented on packages/opentelemetry-instrumentation-pinecone/tests/test_query.py:
```diff
@@ -152,12 +154,12 @@ def test_pinecone_retrieval(traces_exporter, metrics_reader, openai_client):
                         assert data_point.sum > 0
 
-                if metric.name == "db.pinecone.query.scores":
+                if metric.name == f"{SpanAttributes.PINECONE_DB_QUERY}.scores":
```
`METERS.PINECONE_SCORES`

nirga commented on packages/opentelemetry-instrumentation-pinecone/tests/test_query.py:
```diff
@@ -168,7 +170,7 @@ def test_pinecone_retrieval(traces_exporter, metrics_reader, openai_client):
                         )
 
-                if metric.name == "db.pinecone.usage.read_units":
+                if metric.name == f"{SpanAttributes.PINECONE_DB_USAGE}.read_units":
```
`METERS.PINECONE_READ_UNITS`

nirga commented on packages/opentelemetry-instrumentation-qdrant/tests/test_qdrant_instrumentation.py:
```diff
@@ -56,13 +56,13 @@ def test_qdrant_upload_collection(exporter, qdrant):
 
     assert (
-        span.attributes.get("qdrant.upload_collection.collection_name")
+        span.attributes.get(f"{SpanAttributes.QDRANT_UPLOAD_COLLECTION}.collection_name")
```
`SpanAttributes.QDRANT_UPLOAD_COLLECTION_NAME`

nirga reviewed (commented):
Thanks so much @tibor-reiss this is great work that you've done and super helpful! However, we need to actually extract ALL constants (except for specific ones of the prompts and completions), since these are being used by other systems to we need to be explicit about it. I left comments with some examples, but I can also take it from here since I know it's a repetitive work :)

tibor-reiss commented:
Thanks for feedback. I have made the corresponding changes. Couple of questions which came up during these adjustments:

- anthropic: WRAPPED_METHODS -> span_name and metric_name: are always the same. This does not seem correct, is it?

- openai: you removed embeddings_duration_histogram which is connected to "llm.openai.embeddings.duration". However, it's used in the tests. Was it a mistake to remove, or should the tests be updated, please?

- weaviate: This implementation does not fit into the current constant-refactoring. Here an example:
"span_name": f"{SpanAttributes.WEAVIATE_SCHEMA}.get",
_SchemaInstrumentorV3
	namespace = f"{AISpanAttributes.WEAVIATE_SCHEMA}"
	mapped_attributes = {
        "get": ["class_name"],
		...
	}
The tests also just use SpanAttributes.WEAVIATE_SCHEMA.
If one would introduce SpanAttributes.WEAVIATE_SCHEMA_GET in accordance with the current refactoring, then I think the namespace attribute should be dropped, and mapped_attributes updated accordingly.
What do you think?

nirga commented:
> Thanks for feedback. I have made the corresponding changes. Couple of questions which came up during these adjustments:
> 
> * anthropic: `WRAPPED_METHODS` -> `span_name` and `metric_name`: are always the same. This does not seem correct, is it?
You're right. Metric name should be removed because the exception count should just be an attribute of the latency metric. I've fixed the span name.

> * openai: you removed `embeddings_duration_histogram` which is connected to `"llm.openai.embeddings.duration"`. However, it's used in the tests. Was it a mistake to remove, or should the tests be updated, please?
Tests should be updated. Fixed that.

> * weaviate: This implementation does not fit into the current constant-refactoring. Here an example:
>   `"span_name": f"{SpanAttributes.WEAVIATE_SCHEMA}.get",`
> 
> ```
> _SchemaInstrumentorV3
> 	namespace = f"{AISpanAttributes.WEAVIATE_SCHEMA}"
> 	mapped_attributes = {
>         "get": ["class_name"],
> 		...
>  	}
> ```
> 
> The tests also just use `SpanAttributes.WEAVIATE_SCHEMA`. If one would introduce `SpanAttributes.WEAVIATE_SCHEMA_GET` in accordance with the current refactoring, then I think the `namespace` attribute should be dropped, and `mapped_attributes` updated accordingly. What do you think?

Namespace shouldn't use the Span Attributes constant. The Span Attributes constant should only be used for actual attributes we put on spans.

tibor-reiss commented:
@nirga

Fixed the merge conflicts (quite a few) + updated the corresponding spans
Fixed openai tests (and the syntax errors from tuple unpacking seem to sneak back, e.g. L134 in opeani/v1/__init__.py :) )

Still need 2 todos (added in the PR above)

I might be wrong, but the merge conflicts are increasing and getting more and more difficult to resolve. Would it be a good idea to start pushing (some) changes to master?

nirga commented:
@tibor-reiss yep! I'll work on wrapping this PR

tibor-reiss commented on packages/opentelemetry-instrumentation-langchain/tests/test_lcel.py:
```diff
@@ -1,21 +1,33 @@
+from opentelemetry.semconv.ai import SpanAttributes
+
+
+CHAT_OPENAI_LANGCHAIN_TASK = "ChatOpenAI.langchain.task"
```
I understand, makes sense.

tibor-reiss commented:
Thanks for merging! I see there were still some recurring issues - I most probably overlooked it during the resolution of conflicts - apologies for the double work.

nirga commented:
@tibor-reiss no worries it was a complex PR
